### PR TITLE
Add Recent Note List

### PR DESCRIPTION
**Pull request by shvkhzod**
Added Recent Note List
- Added publicId, createdAt and updatedAt fields to entity Note
- Added getNoteListByCreatorId method to note.repository
- Created NoteList entity, NoteList Service , useNoteList and GetNoteListResponsePayload
- Created NoteView component to display notes in list in main page

### CODE REVIEW

File: src/application/services/useNoteList.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import type { Ref } from 'vue';
+import { onMounted, ref, watch } from 'vue';
+import { useAppState } from './useAppState';
+
+/**
+ * Note list hook state
+ */
+interface UseNoteListComposableState {
+  /**
+   * NoteSettings ref
```
Comment (slaveeks) [inline src/application/services/useNoteList.ts:12]:
```suggestion
   * NoteList ref
```

File: src/application/services/useNoteList.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+     * Get user id
+     */
+    const user = useAppState().user.value;
+
+
+    /**
+     * If user is logged in, load note list
+     */
+
+    if (user !== undefined && user !== null) {
```
Comment (slaveeks) [inline src/application/services/useNoteList.ts:47]: maybe simplify it by `if (user)`?

File: src/application/services/useNoteList.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      noteList.value = await noteListService.getNoteListByCreatorId(user.id, 1);
```
Comment (slaveeks) [inline src/application/services/useNoteList.ts:48]: Hardcoded page number, maybe pass it to `load()`?

File: src/domain/entities/Note.ts [hunk 4]
```diff
@@ -14,6 +19,12 @@ export type NoteContent = OutputData;
  * Note entity
  */
 export interface Note {
+
+  /**
+   * Note public id
+   */
+     publicId?: NotePublicId;
```
Comment (slaveeks) [inline src/domain/entities/Note.ts:26]: Why it can be `undefined`, if we have required content?

File: src/domain/entities/Note.ts [hunk 5]
```diff
@@ -23,4 +34,15 @@ export interface Note {
    * Note content
    */
   content: NoteContent;
+
+  /**
+   * When note was created
+   */
+  createdAt?: string;
```
Comment (slaveeks) [inline src/domain/entities/Note.ts:41]: Like  comment about`publicId`

File: src/domain/entities/Note.ts [hunk 6]
```diff
@@ -23,4 +34,15 @@ export interface Note {
...
+
+  /**
+   * Last time note was updated
+   */
+  updatedAt?: string;
```
Comment (slaveeks) [inline src/domain/entities/Note.ts:46]: Like  comment about`publicId`

File: src/domain/noteList.service.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  /**
+   * Returns note list by creator id
+   *
+   * @param userId - user id
+   * @param page - number of current pages
+   * @returns { Promise<NoteList> } list of notes
+   */
+  public async getNoteListByCreatorId(userId: number, page: number): Promise<NoteList> {
+    return {
+      items: (await this.repository.getNoteListByCreatorId(userId, page)).items,
```
Comment (slaveeks) [inline src/domain/noteList.service.ts:28]: maybe `return await ...`?

**Comment by neSpecc** [comment]
Please, fill the PR description for people who want to review it

File: src/domain/entities/Note.ts [hunk 8]
```diff
@@ -14,6 +19,12 @@ export type NoteContent = OutputData;
  * Note entity
  */
 export interface Note {
+
+  /**
+   * Note public id
+   */
+  publicId: NotePublicId;
```
Comment (neSpecc) [inline src/domain/entities/Note.ts:26]: note should now contain `publicId`. For the frontend-side there is only "id" which is public `id`.

Only Backend knows that note have both "public id" and "internal id"

File: src/application/services/useAppState.ts [hunk 9]
```diff
@@ -31,6 +31,7 @@ export const useAppState = createSharedComposable((): UseAppStateComposable => {
     }
   });
 
+
```
Comment (neSpecc) [inline src/application/services/useAppState.ts:34]: please, remove unnecessary changes

File: src/domain/entities/NoteList.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import type { Note } from './Note';
+
+/**
+ * Note list  entity.
+ * An object with the "items" property containing a list of all existing notes created by the user
+ */
+export type NoteList = {
+    items: Note[];
```
Comment (neSpecc) [inline src/domain/entities/NoteList.ts:8]: seems like there is 4 space indentation instead of 2. Install "editorconfig" plugin to your IDE so it will respect the `.editorconfig` file

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: src/domain/entities/Note.ts (hunk 8 above)
Comment (shvkhzod) [reply src/domain/entities/Note.ts:26]: you mean, there should only be publicId? no other ids?

Comment (neSpecc) [reply src/domain/entities/Note.ts:26]: yes, but public id should be named just "id". 

The idea is to completely hide internal "id" from API users. So they could not loop over it and check how many articles we have. So "public" prefix makes sense only for backend. For frontend, there is only "note.id" (which is public id)

File: src/application/services/useNoteList.ts (hunk 2 above)
Comment (TatianaFomina) [reply src/application/services/useNoteList.ts]: bad advise 😞

File: src/domain/note.service.ts [hunk 11]
```diff
@@ -57,6 +58,8 @@ export default class NoteService {
    * @param content - Note content (Editor.js data)
    */
   public async updateNoteContent(id: string, content: NoteContent): Promise<void> {
+    useEventBus('note').emit('note-updated', id);
```
Comment (neSpecc) [inline src/domain/note.service.ts:61]: seems like it is not used anywhere

File: src/application/services/useNoteList.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+
+    /**
+     * If user is logged in, load note list
+     */
+
+    if (user) {
+      noteList.value = await noteListService.getNoteListByCreatorId(user.id, page);
+    } else {
+      console.log('User is not logged');
```
[the remark this answers is not in the document]
Comment (HyTekCoop) [reply src/application/services/useNoteList.ts:54]: remove log

File: src/application/services/useNoteList.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  onMounted(async () => {
+    await load(1);
+  });
+
+  watch(useAppState().user, async () => {
+    await load(1);
+  });
+
+  return {
+
```
[the remark this answers is not in the document]
Comment (HyTekCoop) [reply src/application/services/useNoteList.ts:67]: Delete all unnecessary lines

File: src/domain/event-bus/index.ts [hunk 14]
```diff
@@ -1,6 +1,7 @@
 import type { AUTH_COMPLETED_EVENT_NAME, AuthCompletedEvent } from './events/AuthCompleted';
 import type { AUTH_LOGOUT_EVENT_NAME, AuthLogoutEvent } from './events/AuthLogoutEvent';
 
+
```
[the remark this answers is not in the document]
Comment (HyTekCoop) [reply src/domain/event-bus/index.ts:4]: Delete all unnecessary lines

File: src/infrastructure/transport/notes-api/types/GetNoteListResponsePayload.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+import type { NoteList } from '@/domain/entities/NoteList';
+
+/**
+ * Get note list response payload
+ */
+export type GetNoteListPayload = NoteList;
```
[the remark this answers is not in the document]
Comment (HyTekCoop) [reply src/infrastructure/transport/notes-api/types/GetNoteListResponsePayload.ts:6]: Remove type, use NoteList

File: src/application/services/useNoteList.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+   * NoteList ref
+   */
+  const noteList = ref<NoteList | null>(null);
+
+  /**
+   * Get note list
+   *
+   * @param page - number of pages
+   */
+  const load = async (page:number): Promise<void> => {
```
Comment (neSpecc) [inline src/application/services/useNoteList.ts:38]:
```suggestion
  const load = async (page: number): Promise<void> => {
```

File: src/presentation/pages/Home.vue [hunk 17]
```diff
@@ -1,19 +1,55 @@
...
+      <Input text="Input text" />
+
+      <div v-if="noteList?.items.length">
+        <div
+          v-for="note in noteList.items"
+          :key="note.id"
+        >
+          <NoteView
+            :title="note.content.blocks[0]?.data?.text"
+            :subtitle="note.content.blocks[1]?.data.text || ''"
```
Comment (neSpecc) [inline src/presentation/pages/Home.vue:22]: its better to use computed properties inside NoteView

File: src/application/services/useNoteList.ts [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+export default function (): UseNoteListComposableState {
+  /**
+   * NoteList ref
+   */
+  const noteList = ref<NoteList | null>(null);
+
+  /**
+   * Current page
+   */
+  let currentPage = 1;
```
Comment (neSpecc) [inline src/application/services/useNoteList.ts:41]: you need to clear this variable after leaving a page

Comment (neSpecc) [reply src/application/services/useNoteList.ts:41]: use `onUnmounted` for that

File: src/domain/noteList.service.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  /**
+   * Returns note list by creator id
+   *
+   * @param userId - user id
+   * @param page - number of current pages
+   * @returns { Promise<NoteList> } list of notes
+   */
+  public async getNoteListByCreatorId(userId: number, page: number): Promise<NoteList> {
```
Comment (neSpecc) [inline src/domain/noteList.service.ts:26]: userId is redundant

File: src/presentation/components/note/NoteList.vue [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    <div
+      v-for="note in noteList.items"
+      :key="note.id"
+    >
+      <NoteView
+        :note="note"
+        @click="router.push('/note/' + note.id)"
+      />
+    </div>
+    <button @click="loadMoreNotes">Load more notes</button>
```
Comment (neSpecc) [inline src/presentation/components/note/NoteList.vue:12]: i18n

File: src/presentation/components/note/NoteView.vue
Comment (neSpecc) [inline src/presentation/components/note/NoteView.vue:1]: components/note-list/NoteListItem.vue

File: src/presentation/components/note/NoteList.vue
Comment (neSpecc) [inline src/presentation/components/note/NoteList.vue:1]: components/note-list/NoteList.vue

File: src/presentation/components/note/NoteView.vue [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+  const firstNoteBlock = props.note.content.blocks[0];
+
+  if (!firstNoteBlock || !Boolean(firstNoteBlock.data.text)) {
+    return 'Note';
+  } else {
+    return firstNoteBlock.data.text.slice(0, limitCharsForNoteTitle);
+  }
+});
+
+const noteSubTitle = computed(() => {
```
Comment (neSpecc) [inline src/presentation/components/note/NoteView.vue:41]: jsdocs are missed

File: src/presentation/components/note/NoteView.vue [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+  note: Note;
+}>();
+
+const limitCharsForNoteTitle = 50;
+const limitCharsForNoteSubTitle = 96;
+
+const noteTitle = computed(() => {
+  const firstNoteBlock = props.note.content.blocks[0];
+
+  if (!firstNoteBlock || !Boolean(firstNoteBlock.data.text)) {
```
Comment (neSpecc) [inline src/presentation/components/note/NoteView.vue:34]:
```suggestion
  if (!firstNoteBlock || firstNoteBlock.data.text === undefined) {
```

File: src/presentation/components/note/NoteView.vue [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+    return 'Note';
```
Comment (neSpecc) [inline src/presentation/components/note/NoteView.vue:35]:
```suggestion
    return 'Untitled';
```

File: src/presentation/pages/Home.vue [hunk 24]
```diff
@@ -1,19 +1,27 @@
 <template>
   <div :class="$style.page">
     <h1>{{ $t('home.title') }}</h1>
-    <p>This page will contain your Notes you recently worked with</p>
+    <div v-if="user">
+      <Heading :level="2"> CodeX UI showcase </Heading>
+      <Button text="Button text" />
+      <Input text="Input text" />
```
Comment (neSpecc) [inline src/presentation/pages/Home.vue:7]:
```suggestion
```

File: src/presentation/pages/Home.vue [hunk 25]
```diff
@@ -1,19 +1,27 @@
...
+    <div v-else>
+      <p>{{ $t('home.authText') }}</p>
+    </div>
   </div>
 </template>
 
 <script setup lang="ts">
 import { useHead } from 'unhead';
 import { useI18n } from 'vue-i18n';
 import { Button, Input, Heading } from 'codex-ui/vue';
```
Comment (neSpecc) [inline src/presentation/pages/Home.vue:20]:
```suggestion
```

File: src/presentation/pages/Home.vue [hunk 26]
```diff
@@ -25,8 +33,10 @@ useHead({
 </script>
 
 <style lang="postcss" module>
+@import '@/presentation/styles/typography.pcss';
+
 .page {
-  background-color: var(--ui-color);
+  //background-color: var(--ui-color);
```
Comment (neSpecc) [inline src/presentation/pages/Home.vue:39]:
```suggestion
```

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: src/application/i18n/messages/en.json [hunk 27]
```diff
@@ -26,8 +26,15 @@
   "note": {
     "new": "New Note"
   },
+  "noteList": {
+    "buttons": {
+      "loadMoreNotes": "Load more notes"
+    },
+    "noNoteList": "No Notes yet. Make your first note"
```
Comment (slaveeks) [inline src/application/i18n/messages/en.json:33]:
```suggestion
    "emptyNoteList": "No Notes yet. Make your first note"
```

File: src/domain/note.repository.interface.ts [hunk 28]
```diff
@@ -22,6 +23,14 @@ export default interface NoteRepositoryInterface {
    */
   getNoteByHostname(hostname: string): Promise<{ note: Note; accessRights: NoteAccessRights }>;
 
+  /**
+   * Returns a list of notes by creator id
+   *
+   * @param userId - user id
+   * @param page - number of pages
+   */
+  getNoteListByCreatorId(page: number): Promise<NoteList>;
```
Comment (neSpecc) [inline src/domain/note.repository.interface.ts:32]:
```suggestion
  getNoteList(page: number): Promise<NoteList>;
```

File: src/domain/note.repository.interface.ts [hunk 29]
```diff
@@ -22,6 +23,14 @@ export default interface NoteRepositoryInterface {
    */
   getNoteByHostname(hostname: string): Promise<{ note: Note; accessRights: NoteAccessRights }>;
 
+  /**
+   * Returns a list of notes by creator id
```
Comment (neSpecc) [inline src/domain/note.repository.interface.ts:27]:
```suggestion
   * Returns a user note list
```

File: src/domain/note.repository.interface.ts [hunk 30]
```diff
@@ -22,6 +23,14 @@ export default interface NoteRepositoryInterface {
...
+   *
+   * @param userId - user id
```
Comment (neSpecc) [inline src/domain/note.repository.interface.ts:29]:
```suggestion
```

File: src/domain/noteList.service.ts [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+   * @param repository - note repository
+   */
+  constructor(repository: NoteRepository) {
+    this.repository = repository;
+  }
+
+  /**
+   * Returns note list by creator id
+   *
+   * @param userId - user id
```
Comment (neSpecc) [inline src/domain/noteList.service.ts:22]:
```suggestion
```

File: src/domain/noteList.service.ts [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+   * Note service constructor
+   *
+   * @param repository - note repository
+   */
+  constructor(repository: NoteRepository) {
+    this.repository = repository;
+  }
+
+  /**
+   * Returns note list by creator id
```
Comment (neSpecc) [inline src/domain/noteList.service.ts:20]:
```suggestion
   * Returns user note list
```

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: src/application/services/useNoteList.ts [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+  /**
+   * Get note list
+   *
+   * @param page - number of pages
+   */
+  const load = async (page: number): Promise<void> => {
+    /**
+     * Get user id
+     */
+    const user = useAppState().user.value;
```
Comment (slaveeks) [inline src/application/services/useNoteList.ts:52]: I think, it is redundant check

File: src/presentation/components/note-list/NoteList.vue [hunk 34]
```diff
@@ -0,0 +1,29 @@
...
+      <NoteView
+        :note="note"
+        @click="router.push('/note/' + note.id)"
+      />
+    </div>
+    <button @click="loadMoreNotes">{{ $t('noteList.buttons.loadMoreNotes') }}</button>
+  </div>
+
+  <div v-else>
+    <p>{{ $t('noteList.emptyNoteList') }}</p>
```
Comment (TatianaFomina) [inline src/presentation/components/note-list/NoteList.vue:16]: better use `t` via `useI18n`

like [here](https://github.com/codex-team/notes.web/blob/main/src/presentation/pages/NoteSettings.vue#L43)

File: src/presentation/components/note-list/NoteListItem.vue [hunk 35]
```diff
@@ -0,0 +1,98 @@
...
+const limitCharsForNoteSubTitle = 96;
+
+/**
+ * Get the title from Note
+ */
+const noteTitle = computed(() => {
+  const firstNoteBlock = props.note.content.blocks[0];
+
+  if (!firstNoteBlock || firstNoteBlock.data.text === undefined) {
+    return 'Untitled';
```
Comment (TatianaFomina) [inline src/presentation/components/note-list/NoteListItem.vue:41]: this should also have internationalisation

File: src/presentation/components/note-list/NoteListItem.vue [hunk 36]
```diff
@@ -0,0 +1,98 @@
+<template>
+  <div class="note">
+    <div class="note-header">
```
Comment (TatianaFomina) [inline src/presentation/components/note-list/NoteListItem.vue:3]: try using css modules like here https://github.com/codex-team/notes.web/blob/main/src/presentation/pages/Home.vue#L2

**Review by neSpecc — APPROVED** [review]

**Review by slaveeks — APPROVED** [review]
